### PR TITLE
Added delete-to-next/previous-word-boundary

### PR DESCRIPTION
**Pull request by ghost**
This is better behavior for alt-del/backspace operations, is less "greedy" and doesn't eat up excess characters.
Also set default key mappings to these, feel free to trash my changes to keymaps/darwin.cson. Though I do feel this should be default behavior, personally.

### CODE REVIEW

**Comment by CaptSaltyJack** [comment]
Any update on this?

**Comment by dogancelik** [comment]
Can someone merge this?

**Comment by 50Wliu** [comment]
@CaptSaltyJack could you please rebase this?  There's conflicts right now.

**Comment by CaptSaltyJack** [comment]
@50Wliu My git skills aren't mad enough, I'd probably wind up destroying my repo. Shouldn't I just do `git merge master` to merge updates into the feature branch?

**Comment by CaptSaltyJack** [comment]
http://stackoverflow.com/a/21539012/685125

**Comment by benogle** [comment]
```
git checkout master
git pull
git checkout my-branch
git rebase master
```

then fix all the conflicts

```
git add ... # all the files conflicted
git rebase --continue
```

Then when all the conflicts are fixed

```
git push --force
```

**Comment by CaptSaltyJack** [comment]
Done! `git pull` did nothing since that was my own remote (origin). I had to merge from `upstream/master` first.

**Comment by benogle** [comment]
Oh right, forks.

**Comment by CaptSaltyJack** [comment]
Just curious, why rebase over merge? Is it to eliminate all the history I've created, squashing all those commits into a single one?

**Comment by 50Wliu** [comment]
Not too sure about rebase vs merge, but I've generally always used rebase when I had conflicting changes.

Also, could you please add your changes to win32.json and linux.json if/when the changes to darwin.json get a :+1:?

**Comment by CaptSaltyJack** [comment]
Yeah, as I understand it, rebase wipes out a branch's history. Which I guess is OK when people are merging changes into a project like this. No one needs to see my 20 revisions saying "fixed that one thing again." All that matters is the single commit with one comment as it's merged into upstream/master.

File: spec/text-editor-spec.coffee [hunk 1]
```diff
@@ -2343,6 +2343,61 @@ describe "TextEditor", ->
...
+          editor.deleteToPreviousWordBoundary()
+          expect(buffer.lineForRow(1)).toBe 'var quicksort function () {'
+          expect(buffer.lineForRow(2)).toBe '  var sort =(items) {'
+          expect(cursor1.getBufferPosition()).toEqual [0, 14]
+          expect(cursor2.getBufferPosition()).toEqual [1, 12]
+
+      describe "when text is selected", ->
+        it "deletes only selected text", ->
+          editor.setSelectedBufferRange([[1, 24], [1, 27]])
+          editor.deleteToEndOfWord()
```
Comment (lee-dohm) [inline spec/text-editor-spec.coffee]: Is this supposed to be `editor.deleteToPreviousWordBoundary()`?

File: spec/text-editor-spec.coffee [hunk 2]
```diff
@@ -2343,6 +2343,61 @@ describe "TextEditor", ->
...
+          editor.deleteToNextWordBoundary()
+          expect(buffer.lineForRow(1)).toBe 'var quicksort = () {'
+          expect(buffer.lineForRow(2)).toBe '  var sort = function(it {'
+          expect(cursor1.getBufferPosition()).toEqual [1, 24]
+          expect(cursor2.getBufferPosition()).toEqual [2, 5]
+
+      describe "when text is selected", ->
+        it "deletes only selected text", ->
+          editor.setSelectedBufferRange([[1, 24], [1, 27]])
+          editor.deleteToEndOfWord()
```
Comment (lee-dohm) [inline spec/text-editor-spec.coffee]: And I suspect this is supposed to be `editor.deleteToNextWordBoundary()`?

**Comment by benogle** [comment]
Rebasing does not create unnecessary merge commits. Basically its as if you just developed this on the latest version of master. Your history is still there, but the commit shas are different.

File: spec/text-editor-spec.coffee (hunk 1 above)
Comment (CaptSaltyJack) [inline spec/text-editor-spec.coffee]: @lee-dohm Oops.. yes, probably. I don't have the test text handy so I can't verify, but it doesn't make any sense that the test doesn't even call the new function.

File: spec/text-editor-spec.coffee (hunk 2 above)
Comment (CaptSaltyJack) [inline spec/text-editor-spec.coffee]: @lee-dohm Yes. I'll fix both of these.

**Comment by lee-dohm** [comment]
I ran the specs from this PR with `apm test` on Atom change adb0178 on Mac OS X 10.10.3 and both "when no text is selected" tests are failing.

**Comment by CaptSaltyJack** [comment]
I could never test this properly because of console errors within Atom when running the (if I recall the name) Package Specs command. To clarify, I *tested* this thoroughly myself, but the test specs I was never able to verify. @benogle is familiar with the huge headache I went through to try to get the test specs working. I gave up. I still can't run `atom -d` without errors.

> Uncaught TypeError: Cannot read property 'dispatch' of /Users/bob/github/atom/src/space-pen-extension.coffee:64

**Comment by lee-dohm** [comment]
Ok, I'll take a look and fix the tests.

**Comment by CaptSaltyJack** [comment]
Even `apm test` bombs out:

> [17216:0429/172638:INFO:CONSOLE(27)] "Error: Module version mismatch. Expected 43, got 41.

It would be cool to have an atom developer's guide somewhere that has clear steps on how to set up a dev environment. I could never get mine 100% functional.

**Comment by CaptSaltyJack** [comment]
Thanks @lee-dohm. I'm sure it's something simple/dumb.

**Comment by lee-dohm** [comment]
You're most welcome @CaptSaltyJack :smiley:

**Comment by benogle** [comment]
I am not comfortable changing the default keybinding. I say we merge this without the default bindings changed, and have some folks change their keybindings before changing the default.

**Comment by CaptSaltyJack** [comment]
@benogle That makes sense to me. Probably best not to change default behaviors after people are already used to what's been the default since the beginning.

**Comment by lee-dohm** [comment]
@CaptSaltyJack, I have a question about the third clause to ".deleteToNextWordBoundary() when no text is selected deletes to the next word boundary", starting on line 2384 which is:

```coffee
          editor.deleteToNextWordBoundary()
          expect(buffer.lineForRow(0)).toBe 'var quicksort = () {'
          expect(buffer.lineForRow(1)).toBe '  var sort = function(it {'
          expect(cursor1.getBufferPosition()).toEqual [0, 15]
          expect(cursor2.getBufferPosition()).toEqual [1, 24]
```

I corrected the typos and mismatches (row numbers and cursor positions), but the question is that before this portion of the test is executed, the text for line 0 should be `var quicksort =|function () {` with the pipe character representing the position of the cursor. My understanding is that the intent of this PR is that it is *not* supposed to be greedy about taking whitespace too. Your test states that the result should be `var quicksort =| () {`, but the actual result is `var quicksort =|() {`. Is this a bug?

**Comment by lee-dohm** [comment]
Pushing my updates to fix most of the tests.

**Comment by lee-dohm** [comment]
Looking to see if I can fix it.

**Comment by CaptSaltyJack** [comment]
Geez, was I drunk when I wrote this test? Pushing..

**Comment by CaptSaltyJack** [comment]
Unfortunately I can't even test my own code now, as I can't run Atom in dev mode. @benogle, what were the commands to reset the atom dev environment over again? Something about clean, bootstrap, build?

**Comment by lee-dohm** [comment]
I think you're talking about this command @CaptSaltyJack? `apm/node_modules/atom-package-manager/bin/apm rebuild`

**Comment by CaptSaltyJack** [comment]
@lee-dohm Maybe. I'm rebuilding now to see if it fixes `atom -d` so I can double-check the functionality of my changes.

**Comment by lee-dohm** [comment]
Merged as #6575. Thanks @CaptSaltyJack :+1:

**Pull request closed without merging**